low-level/files/opt/seagate/sspl/setup/resource_map/server.py

```diff
@@ -31,19 +41,169 @@ class ServerMap(ResourceMap):
     name = "server"
 
     def __init__(self):
-        """Initialize server"""
-        pass
+        """Initialize server."""
+        super().__init__()
+        self.server = {
+            'sas_hba': self.get_sas_hba_info,
+            'sas_ports': self.get_sas_ports_info,
+            'nw_ports': self.get_nw_ports_info
+        }
+        self.sysfs = ToolFactory().get_instance('sysfs')
+        self.sysfs.initialize()
 
     def get_health_info(self, rpath):
         """
-        Get health information of fru in given resource id
+        Get health information of fru in given resource id.
 
         rpath: Resource id (Example: nodes[0]>compute[0]>hw>disks)
         """
-        pass
+        info = {}
+        fru_found = False
+        nodes = rpath.strip().split(">")
+        leaf_node, _ = self.get_node_details(nodes[-1])
+        if leaf_node == "compute":
+            for fru in self.server:
+                info.update({fru: self.server[fru]()})
+            info["last_updated"] = int(time.time())
+            fru_found = True
+        else:
+            fru = None
+            for node in nodes:
+                fru, _ = self.get_node_details(node)
+                if self.server.get(fru):
+                    fru_found = True
+                    info = self.server[fru]()
+                    break
+        if not fru_found:
+            raise ResourceMapError(
+                errno.EINVAL,
+                "Health provider doesn't have support for'{rpath}'.")
+        return info
+
+    @staticmethod
+    def get_node_details(node):
+        """
+        Parse node information and returns left string and instance.
+
+        Example
+            "storage"    -> ("storage", "*")
+            "storage[0]" -> ("storage", "0")
+        """
+        res = re.search(r"(\w+)\[([\d]+)\]|(\w+)", node)
+        inst = res.groups()[1] if res.groups()[1] else "*"
+        node = res.groups()[0] if res.groups()[1] else res.groups()[2]
+        return node, inst
+
+    def get_sas_hba_info(self):
+        """Return the latest SAS-HBA information."""
+        sas_hba_data = []
+        hosts = self.sysfs.get_sas_host_list()  # ['host1']
+        for host in hosts:
+            host_id = SAS_RESOURCE_ID + host.replace('host', '')
+            host_data = self.get_data_template(host_id, False)
+            ports = self.sysfs.get_sas_port_list(host)
+            # ports = ['port-1:0', 'port-1:1', 'port-1:2', 'port-1:3']
+            health = "OK"
+            specifics = {
+                'num_ports': len(ports),
+                'ports': []
+            }
+            for port in ports:
+                port_data = self.sysfs.get_sas_port_data(port)
+                specifics['ports'].append(port_data)
+                if port_data['state'] != 'running':
+                    health = "Failed"
```

Comment:
if port is disabled, we cant determine its health, so it should be NA

Reply:
Also currently supported health values - OK, Fault, Degraded, NA

Reply:
Okay, will correct it to NA in new PR.